### DISCUSSION TITLE
Prevent inactive state

### DISCUSSION THREAD

**Discussion by loncar** [TanStack/query#1193] [Q&A] [answered]
Is there a way of preventing inactive state for Query? I have noticed that only observable queries are non inactive. But if I create Query via `queryCache.fetchQuery` I don't have any observers on it and it is inactive right away even if I put `staleTime` and `cacheTime` to infinite. 

Use case that I have is: getting some complex data by fetching across multiple endpoints with multiple requests per endpoint. Logic behind that is in Service part of App that is only at the end of processing used in React Component. But data will be there always and be invalidated by hand. 

So question would be is there a easy way of using QueryCache fetching without Hooks and have caching layer with data being active (no inactive state)? 

Thanks :)

**Comment by TkDodo** [comment, marked as answer]
So why would you need it to be in `active` state? As you said, only queries that have observers (rendered components via `useQuery` etc) are `active`. A query that is `inactive` will be garbage-collected after the `cacheTime` is reached - defaults to 5 minutes. If I got this right and you are only fetching, but never rendering, you can just increase the `cacheTime` and you data will be available via `queryCache.getQueryData`.

**Reply by loncar** [reply]
Reason why I want them to be in active state is to avoid new calls to server. If that cached data is in Active state then react query will serve them instead of calling the backend again. Also one reason for that is complex data structure that is requiring processing before it is display to user. For example if user make some changes on filtering those records then I will be able to reuse like 70% of data that is already there. Using `queryCache.getQueryData`  will be hard at that point because I would have to track what is there and what is missing.

**Reply by TkDodo** [reply]
So, I still don't think I fully understand the use-case:

- you put some data in the cache with `queryCache.fetchQuery` - it will become inactive because there are no observers.
- if you increase the `cacheTime` (say, to `Infinity`), it will be there forever.
- if you read it via `queryCache.getQueryData`, you should just get the data from the cache without re-fetching.

so, if you are not using `useQuery` but you are also not using `queryCache.getQueryData`, how are you reading the data from the cache...?

**Reply by loncar** [reply]
I will call each time `queryCache.fetchQuery` when I want to get the data, but I don't want each time to make new call to server. I am using same `queryCache` object for that. If data were active (fresh) at that point of time new call wouldn't be made.  

If I were to use `queryCache.getQueryData` then I would need first to check if there is any data and then if there is none to make `fetchQuery` call. That is something that I want to avoid by simply running `fetchQuery`. I want to have simple caching layer that will sit in between without need to check if there is already data there or I need to grab new one etc.

Story behind this case that I have: need to fetch data from 6 different data sources and there will be at least 2 calls per source. So min 12 calls but it can be like 6 calls instead of 2 per source that will end up in 36 requests. It all depends on the users selection. Data don't need to be fresh so I have used staleTime but that don't count if I have inactive state of data :(

**Reply by TkDodo** [reply]
Ah, now I got it, thanks for explanation. I was confusing `fetchQuery` with `prefetchQuery`.

FYI, `fetchQuery` does respect `staleTime`, however, there is a bug in v2 where it does not respect the global one (just the one from the options you pass to it), which has been fixed in v3.

So if you do: `queryCache.fetchQuery('key', fetchFn, { staleTime: Infinity })` I think it should just fetch once, no matter if it's active or inactive.

**Reply by loncar** [reply]
Thanks, that was it. I was using 2.* version and `staleTime` prop on `QueryCache` obj instead of setting it on each query. Thanks a lot. :)
Setting it on each query is working fine, I was also looking at 3. version but API looks different so it would require some changes but lack of docs prevented me from doing that atm.

**Reply by TkDodo** [reply]
have a look here for upgrading info: https://react-query-beta.tanstack.com/guides/migrating-to-react-query-3

**Reply by loncar** [reply]
Thanks, I will follow that doc :)

**Reply by ivansolobear-omni** [reply]
hey , I currently face a similiar problem but I use V3 already:

I have a query, which uses an object of query params as keys. I use staleTime: infinity, but what I get is upon changing the params the query becomes inactive. which once I get cached keys I expect to show the cached keys as its suppose to be fresh cashed data. but that Is not what happens and instead he refetches the data every time.

So example:

I have key {from: x, to: y}
I change key to {from: x, to: z}
now i have 2 records cached. 
{from: x, to: z} is currently fresh.
but
{from: x, to: y} is inactive.
and so once I change it back to {from: x, to: y}
I expect it to use the cached data, meaning switching directly from inactive to fresh.
What really happens is it becomes in a refetching state and refetch data from server.

mind you that the keys change using input fields, and the components do not unmount, so maybe I don't use it correctly or something. but I could not find a solution in the documentation.

**Reply by TkDodo** [reply]
this should work. can you show a codesandbox reproduction?

**Reply by sathish-karunakaran** [reply]
Same issue for me too.. using v4

**Reply by sathish-karunakaran** [reply]
![image](https://github.com/TanStack/query/assets/93376354/80c009a8-8fc2-428a-97e1-dc9f4cba47e3)

**Reply by sathish-karunakaran** [reply]
If you see 3rd,4th row items are forcefully goes to inactive state. but it should be on stale state. looking forward the solution

**Reply by TkDodo** [reply]
@sathish-karunakaran 

1) a screenshot is not a codesandbox reproduction!
2) all I see is that you have a query that listens to the key for `get-porfolio-asset-scores` with an id that probably changes over time, going from `4` to `5` and then `55`. As you likely only have have one `useQuery` call that looks something like this:

```
useQuery({
  queryKey: ['portfolio-service/get-portfolio-asset-scores', id, "AAPL US"],
  queryFn,
})
```

the output is totally expected, as only one `id` is active at the same time. This is the most I can tell from just a screenshot

**Reply by sathish-karunakaran** [reply]
Hi @TkDodo 

Thanks for the reply

1. Try to reproduction on codesandbox with different API. since its not available for public
2. Yes I am using same usequery and changing the id field based on the user navigation. 

navigating from `4` to `5`  -> 5 will be stale and 4 will be inactive
Again I will navigating from `5` to `4`  -> it needs to taken from the cache data but instead its fetching from service.

**Reply by TkDodo** [reply]
> navigating from 4 to 5 -> 5 will be stale and 4 will be inactive

that is expected.


> Again I will navigating from 5 to 4 -> it needs to taken from the cache data but instead its fetching from service.

with the default `staleTime` of zero, it will give you data from the cache immediately AND do a background refetch at the same time. If you don't want that, set a higher `staleTime`.

**Reply by bombillazo** [reply]
> with the default `staleTime` of zero, it will give you data from the cache immediately AND do a background refetch at the same time. If you don't want that, set a higher `staleTime`.

Is there a way to not background refetch and use the same cache data until the `staleTime` has elapsed?

**Reply by TkDodo** [reply]
> Is there a way to not background refetch and use the same cache data until the staleTime has elapsed?

That's exactly what staleTime does so not sure I understand the question...
